File: Stardate.cs
﻿using Microsoft.VisualBasic;
using System;
using System.Drawing;
using System.Numerics;
using System.Runtime.InteropServices;
using System.Security.Cryptography;
using System.Security.Cryptography.Xml;
using System.Xaml;
using System.Xml.Schema;

public class Stardate
{
    /************************************************************

    TOS    Starts  2266  0193.29  +1496.5725
           Ends    2270  6179.58    
    TMP    Starts  2267  5701.73  +133.07789
           Ends    2286  8230.21
    Films  Starts  2283  7697.87  +188.116
           Ends    2294  9767.15
    TNG    Starts  2323  0.0      +1000.0


    Some key stardates for reference
    --------------------------------

    1512.0    2266    The Corbomite Maneuvre
    5693.0    2269    Tholian Web, assume that this is 4 years on

    7412.0    2279    Star Trek: The Motion Picture
    8130.0    2285    The Wrath of Khan
    8210.0    2285    The Search for Spock
    8390.0    2286    The Voyage Home 
    8454.0    2287    The Final Frontier
    9521.0    2292    The Undiscovered country
    9715.0    2293    Generations

    54868.6   23--    868.6 along the year is a few days after First Contact Day.

    ************************************************************/

    private const double TOS_ROOT = 2265.1893d; // Changed from 2265.8709 for aesthetic reasons
    private const double TOS_INCREMENT = 1496.2162d; // Changed from 1496.5725d
    private const double TMP_ROOT = 2224.155d;
    private const double TMP_INCREMENT = 133.07789d;
    private const double FILMS_ROOT = 2242.08d;
    private const double FILMS_INCREMENT = 188.116d;
    private const double TNG_ROOT = 2323.3981d; // __868.6 should be a few days after First Contact day (April 5th).
    private const double TNG_INCREMENT = 1000d;

    private DateTime earthDate;
    private double tosMetric;
    private double tmpMetric;
    private double filmsMetric;
    private double tngMetric;

    public string FormatStardate(object stardate)
    {
        if (stardate is DateTime dt)
            return dt.ToString("dd/MM/yyyy HH:mm");
        if (stardate is double d)
            return ((int)(d * 100d) / 100d).ToString();
        else
            return "Error";
    }

    public DateTime EarthDate { get { return earthDate; } }
    public double TOS { get { return tosMetric; } }
    public double TMP { get { return tmpMetric; } }
    public double Films { get { return filmsMetric; } }
    public double TNG { get { return tngMetric; } }
    public object Primary
    {
        get
        {
            if (tosMetric < 0)
                return earthDate;
            if (earthDate.Year < 2280)
                return tosMetric < tmpMetric ? tosMetric : tmpMetric;
            if (earthDate.Year < 2300)
                return tmpMetric > filmsMetric ? tmpMetric : filmsMetric;
            return filmsMetric > tngMetric ? filmsMetric : tngMetric;
        }
    }

    public Stardate(DateTime dateTime)
    {
        Calculate(dateTime);
    }
    public Stardate(double stardate)
    {
        // Truncation somewhere in the back and forth of things can lead to stardates of, say, 5000 getting
        // read back as 4999.99.
        stardate += .0000001d;
        if (stardate < 7000)
        {
            double realTOS = (stardate / TOS_INCREMENT) + TOS_ROOT;
            double realTMP = (stardate / TMP_INCREMENT) + TMP_ROOT;
            Calculate(realTOS > realTMP ? realTOS : realTMP);
        }
        else if (stardate < 10000)
        {
            double realTMP = (stardate / TMP_INCREMENT) + TMP_ROOT;
            double realFilms = (stardate / FILMS_INCREMENT) + FILMS_ROOT;
            Calculate(realTMP < realFilms ? realTMP : realFilms);
        }
        else
        {
            double realFilms = (stardate / FILMS_INCREMENT) + FILMS_ROOT;
            double realTNG = (stardate / TNG_INCREMENT) + TNG_ROOT;
            Calculate(realFilms < realTNG ? realFilms : realTNG);
        }
    }

    public void Calculate(double dateReal)
    {
        // Convert the double to a DateTime.
        DateTime dt = new DateTime((int)dateReal, 1, 1);
        Calculate(dateReal, dt.AddMilliseconds((DateTime.IsLeapYear(dt.Year) ? 31_622_400_000 : 31_536_000_000) *
                                               (dateReal % 1)));
    }
    public void Calculate(DateTime dateTime)
    {
        // generate and return a double representing the year and fraction through the year.
        double secondsAlongYear = ((dateTime.DayOfYear - 1) * 86_400d) +
                                  (dateTime.Hour * 3_600) +
                                  (dateTime.Minute * 60) +
                                  (dateTime.Second) +
                                  (dateTime.Millisecond / 1000d);
        Calculate(dateTime.Year + (secondsAlongYear / (DateTime.IsLeapYear(dateTime.Year) ? 31_622_400 : 31_536_000)),
                  dateTime);
    }
    public void Calculate(double dateReal, DateTime dateTime)
    {
        earthDate = dateTime;
        tosMetric = (dateReal - TOS_ROOT) * TOS_INCREMENT;
        tmpMetric = (dateReal - TMP_ROOT) * TMP_INCREMENT;
        filmsMetric = (dateReal - FILMS_ROOT) * FILMS_INCREMENT;
        tngMetric = (dateReal - TNG_ROOT) * TNG_INCREMENT;
    }
}

File: Stardate.js
class Stardate {
    #TOS_ROOT = 2265.1893; // Changed from 2265.8709 for aesthetic reasons
    #TOS_INCREMENT = 1496.2162; // Changed from 1496.5725d
    #TMP_ROOT = 2224.155;
    #TMP_INCREMENT = 133.07789;
    #FILMS_ROOT = 2242.08;
    #FILMS_INCREMENT = 188.116;
    #TNG_ROOT = 2323.3981; // __868.6 should be a few days after First Contact day (April 5th).
    #TNG_INCREMENT = 1000;

    #earthDate = null;
    #tosMetric = null;
    #tmpMetric = null;
    #filmsMetric = null;
    #tngMetric = null;
    
    #IsLeapYear(year) {
        return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
    }

    FormatStardate(stardate) {
        if (stardate instanceof Date) {
            let str = String(stardate.getDate()).padStart(2, '0')
            str += "/" + String(stardate.getMonth() + 1).padStart(2, '0')
            if (stardate.getFullYear() >= 0)
                str += "/" + String(stardate.getFullYear()).padStart(4, '0')
            else
                str += "/-" + String(Math.abs(stardate.getFullYear())).padStart(4, '0')
            str += " " + String(stardate.getHours()).padStart(2, '0')
            str += ":" + String(stardate.getMinutes()).padStart(2, '0')
            return str;
        }            
        if (typeof(stardate) === 'number')
            return String(stardate.toFixed(2));
        else
            return "Error";
    }

    EarthDate() { return this.#earthDate; }
    TOS() { return this.#tosMetric; }
    TMP() { return this.#tmpMetric; }
    Films() { return this.#filmsMetric; }
    TNG() { return this.#tngMetric; }
    Primary(forceStardate) {
        if (this.#earthDate == null)
            return null;
        if (this.#tosMetric < 0 && !forceStardate)
            return this.#earthDate;
        if (this.#tosMetric < 22160) // ~year 2080
            return this.#tosMetric < this.#tmpMetric ? this.#tosMetric : this.#tmpMetric;
        if (this.#tosMetric < 52084) // ~year 2300
            return this.#tmpMetric > this.#filmsMetric ? this.#tmpMetric : this.#filmsMetric;
        return this.#filmsMetric > this.#tngMetric ? this.#filmsMetric : this.#tngMetric;
    }

    Calculate(stardate)
    {
        if (typeof(stardate) === 'number') {
            // Truncation somewhere in the back and forth of things can lead to stardates of, say, 5000 getting
            // read back as 4999.99.
            stardate += .0000001;
            // Convert the stardate value into a number that represents the date in terms of year and fraction of year.
            if (stardate < 7000) {
                let realTOS = (stardate / this.#TOS_INCREMENT) + this.#TOS_ROOT;
                let realTMP = (stardate / this.#TMP_INCREMENT) + this.#TMP_ROOT;
                stardate = realTOS > realTMP ? realTOS : realTMP;
            }
            else if (stardate < 10000) {
                let realTMP = (stardate / this.#TMP_INCREMENT) + this.#TMP_ROOT;
                let realFilms = (stardate / this.#FILMS_INCREMENT) + this.#FILMS_ROOT;
                stardate = realTMP < realFilms ? realTMP : realFilms;
            }
            else {
                let realFilms = (stardate / this.#FILMS_INCREMENT) + this.#FILMS_ROOT;
                let realTNG = (stardate / this.#TNG_INCREMENT) + this.#TNG_ROOT;
                stardate = realFilms < realTNG ? realFilms : realTNG;
            }
            // Convert the double to a DateTime.
            let dt = new Date(Math.trunc(stardate), 0, 1);
            dt.setTime(dt.getTime() +
                       (this.#IsLeapYear(dt.getFullYear()) ? 31622400000 : 31536000000) * (stardate % 1));
            this.#Apply(stardate, dt);
        }
        else if (stardate instanceof Date) {
            // generate and return a double representing the year and fraction through the year.
            let secondsAlongYear = (stardate - new Date(stardate.getFullYear(), 0, 1)) / 1000;
            this.#Apply(stardate.getFullYear() + (secondsAlongYear /
                                                 (this.#IsLeapYear(stardate.getFullYear()) ? 31622400 : 31536000)),
                        stardate);
        }
        else {
            console.error("Invalid input: stardate must be a number or a Date object.");
        }
    }
    #Apply(dateReal, dateTime) {
        if (typeof(dateReal) != 'number' && !(dateTime instanceof Date)) {
            console.error("Invalid input: stardate must be a number or a Date object.");
        } else {
            this.#earthDate = dateTime;
            this.#tosMetric = (dateReal - this.#TOS_ROOT) * this.#TOS_INCREMENT;
            this.#tmpMetric = (dateReal - this.#TMP_ROOT) * this.#TMP_INCREMENT;
            this.#filmsMetric = (dateReal - this.#FILMS_ROOT) * this.#FILMS_INCREMENT;
            this.#tngMetric = (dateReal - this.#TNG_ROOT) * this.#TNG_INCREMENT;
        }
    }
}
